ssg/entities/common.py

```diff
@@ -22,7 +22,7 @@
 def make_items_product_specific(items_dict, product_suffix, allow_overwrites=False):
     new_items = dict()
     for full_label, value in items_dict.items():
-        if "@" not in full_label and full_label not in new_items:
```

Comment:
This is a modification of condition without any compensation, so no wonder that it may change the behavior of the function, as reported in #9894
`new_items` is not an empty dict for the whole time, but the `items_dict` ordering is different between Python2 and Python3, so the loop may behave differently in those interpreters.

Reply:
And it is the case in that issue - `pkgname@rhel7: pam_pkcs11` is supposed to take precedence over `pkgname: openssl-pkcs11`. What happened probably was that `new_items["pkgname"]` was set to `pam_pkcs11`, and it got overwritten by `openssl-pkcs11`. The condition protected product-qualified items from being overwritten by generic items.

Reply:
It feels like this case should fall right into https://github.com/evgenyz/content/blob/5c7cf430a1012bf3151872519f7873b99218f5cb/ssg/entities/common.py#L46 instead of being short-circuited.

Reply:
Otherwise this no-overwrite policy depends on the order of dictionary items (which is not ordered in any way by definition).

Reply:
The funny thing is that CodeClimate was right about that method being too cognitively complex. I think (may be wrong though) that
```
        if not full_label.endswith(product_suffix):
            continue
```
is triggered in case that there is an override attempt of a product-qualified value by a generic value, so now it looks like to me that the exception is not thrown, but the value also isn't overwritten, so it works fine on Python3.
Anyway, only exhaustive unit-testing could bring sufficient level of confidence, and although we would like that, it's out of scope of this PR. The truth is that removing the second part of that conditional was a mistake, and the best thing to do is to fix it - we all know that anybody can make mistakes, and this one was a very educational one that plenty of people can learn from.

Reply:
The truth is that it ain't over yet: https://github.com/ComplianceAsCode/content/pull/9910

Reply:
```
if not full_label.endswith(product_suffix):
            continue
```
happens *after* the possible exception.

Python 2 and 3 difference is in the order of items storage/emission. I've reproduced it in a test case using update().

Reply:
I don't think that the exception happened after this sequence. FTR, the original code can be viewed here: https://github.com/ComplianceAsCode/content/blob/219ea259a4a648d55b4d42380e0853e60a01be79/ssg/build_yaml.py#L949
The issue #9894 was about product-qualified *package*, which is definitely not a "global reference". Therefore, the next line `if label in items_dict and not allow_overwrites and value != items_dict[label]` was one that could trigger an exception, but I don't recall our policy regarding `allow_overwrites`. That clause probably aims to enforce existence of either only product-qualified specs, or no product-qualified specs at all.
Naming sucks, but at least the commit messages that explain why the function got into that shape are quite solid.

Reply:
This function was not used for de-qualifying template variables before refactoring.